File: apps/android/app/src/main/java/com/galaxyssi/chat/voice/VoiceInteractionCommand.kt
package com.galaxyssi.chat.voice

sealed interface VoiceInteractionCommand {
    val sessionId: String
    val idempotencyKey: String

    data class RouteFinalTranscript(
        override val sessionId: String,
        val transcript: TranscriptHypothesis,
        override val idempotencyKey: String
    ) : VoiceInteractionCommand

    data class CancelLegacyWork(
        override val sessionId: String,
        val reasonCode: String,
        override val idempotencyKey: String
    ) : VoiceInteractionCommand
}

data class VoiceInteractionTransition(
    val previous: VoiceInteractionState,
    val current: VoiceInteractionState,
    val commands: List<VoiceInteractionCommand> = emptyList(),
    val accepted: Boolean = true
)

File: apps/ios/GalaxySSI/VoiceInteractionCommand.swift
import Foundation

enum VoiceInteractionCommand: Equatable {
  case routeFinalTranscript(
    sessionId: String,
    transcript: TranscriptHypothesis,
    idempotencyKey: String
  )
  case cancelLegacyWork(
    sessionId: String,
    reasonCode: String,
    idempotencyKey: String
  )

  var sessionId: String {
    switch self {
    case let .routeFinalTranscript(sessionId: sessionId, transcript: _, idempotencyKey: _),
         let .cancelLegacyWork(sessionId: sessionId, reasonCode: _, idempotencyKey: _):
      return sessionId
    }
  }

  var idempotencyKey: String {
    switch self {
    case let .routeFinalTranscript(sessionId: _, transcript: _, idempotencyKey: idempotencyKey),
         let .cancelLegacyWork(sessionId: _, reasonCode: _, idempotencyKey: idempotencyKey):
      return idempotencyKey
    }
  }
}

struct VoiceInteractionTransition: Equatable {
  var previous: VoiceInteractionState
  var current: VoiceInteractionState
  var commands: [VoiceInteractionCommand]
  var accepted: Bool

  init(
    previous: VoiceInteractionState,
    current: VoiceInteractionState,
    commands: [VoiceInteractionCommand] = [],
    accepted: Bool = true
  ) {
    self.previous = previous
    self.current = current
    self.commands = commands
    self.accepted = accepted
  }
}
